src/Compilers/CSharp/Test/Syntax/Parsing/FileModifierParsingTests.cs

```diff
@@ -0,0 +1,1821 @@
+        }
+
+        [Theory]
+        [InlineData(SyntaxKind.ClassKeyword)]
+        [InlineData(SyntaxKind.StructKeyword)]
+        [InlineData(SyntaxKind.InterfaceKeyword)]
+        public void FileModifier_03(SyntaxKind typeKeyword)
+        {
+            UsingNode($$"""
+                partial file {{SyntaxFacts.GetText(typeKeyword)}} C { }
+                """,
+                expectedParsingDiagnostics: new[]
+                {
+                    // (1,14): error CS1002: ; expected
+                    // partial file {{SyntaxFacts.GetText(typeKeyword)}} C { }
+                    Diagnostic(ErrorCode.ERR_SemicolonExpected, SyntaxFacts.GetText(typeKeyword)).WithLocation(1, 14)
+                },
+                expectedBindingDiagnostics: new[]
+                {
+                    // (1,1): error CS0246: The type or namespace name 'partial' could not be found (are you missing a using directive or an assembly reference?)
+                    // partial file interface C { }
+                    Diagnostic(ErrorCode.ERR_SingleTypeNameNotFound, "partial").WithArguments("partial").WithLocation(1, 1),
+                    // (1,9): warning CS0168: The variable 'file' is declared but never used
+                    // partial file interface C { }
+                    Diagnostic(ErrorCode.WRN_UnreferencedVar, "file").WithArguments("file").WithLocation(1, 9),
+                    // (1,14): error CS1002: ; expected
+                    // partial file {{SyntaxFacts.GetText(typeKeyword)}} C { }
+                    Diagnostic(ErrorCode.ERR_SemicolonExpected, SyntaxFacts.GetText(typeKeyword)).WithLocation(1, 14)
+                });
+            N(SyntaxKind.CompilationUnit);
+            {
+                N(SyntaxKind.GlobalStatement);
+                {
+                    N(SyntaxKind.LocalDeclarationStatement);
+                    {
+                        N(SyntaxKind.VariableDeclaration);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken, "partial");
+                            }
+                            N(SyntaxKind.VariableDeclarator);
+                            {
+                                N(SyntaxKind.IdentifierToken, "file");
+                            }
+                        }
+                        M(SyntaxKind.SemicolonToken);
+                    }
+                }
+                N(SyntaxFacts.GetBaseTypeDeclarationKind(typeKeyword));
+                {
+                    N(typeKeyword);
+                    N(SyntaxKind.IdentifierToken, "C");
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+                N(SyntaxKind.EndOfFileToken);
+            }
+            EOF();
+        }
+
+        // PROTOTYPE(ft): is it fine that records parse as a single declaration here, but not other type kinds?
```

Comment:
No, I would expect all of the type kinds to parse like this. What is causing them to parse as above?

Reply:
When we try to parse the partial modifier, we check `IsPartialMember()`:
https://github.com/dotnet/roslyn/blob/228845e4c1fe96cb697f9980285baabbf3da0c53/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L1189-L1192

`IsPartialMember` simply asks, is this `partial`, followed by a possible type, followed by a possible member name.

https://github.com/dotnet/roslyn/blob/228845e4c1fe96cb697f9980285baabbf3da0c53/src/Compilers/CSharp/Portable/Parser/LanguageParser.cs#L1468-L1475

In the case of `partial file record`, this is true. In the case of `partial file class` it is not. So in the former case we parse `partial` as a modifier and go from there to produce a single type declaration, while in the latter case we assume `partial` must be an identifier, and use it as a field type.

This is comparable to the parse we give for `partial async class X`. [SharpLab](https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLQDs4FsIIJYDGqe6MADjAgDQwhTIxUAmIA1AD4ACATAIwCwAKCFkoCGHigAbAAScAHHO4yAGjIDeQgL5CRYidLmKkBAPYImqjdqA=).

My preference here would be to file a "design debt" item to improve error recovery for contextual modifiers and order-sensitive modifiers across the board, and call this parse good enough for the purposes of feature development.